runtime/signals.c

```diff
@@ -287,6 +287,9 @@ void caml_request_minor_gc (void)
   caml_set_action_pending();
 }
 
+// caml_process_pending_callbacks is only called after having taken
+// care of calling caml_gc_dispatch (e.g. via check_urgent_gc), or
+// making sure it is not needed.
```

Comment:
That's indeed a fix to the issue I mentionned. I am not sure that's the simplest to understand, and it breaks the invariant that async callbacks are always called from the same entry point. Also, I don't see why we should call signal handlers when releasing the runtime lock but not other async callbacks.

But I don't think it's too much of an issue, so I'm ok with leaving this as-is.

Reply:
By the way, I tried to do without it, but I think it's the simplest to address the complexity of the situation as witnessed by the length of these PR discussions!